### PR TITLE
[nikobus] add support for Nikobus addresses as seen in the Niko PC application

### PR DESCRIPTION
**Pull request by gitwouter**
Added support for Nikobus addresses as seen in the Niko PC application. In current version only addresses sniffed from the bus can be used which is harder to do for novice users.

Download snapshot here: http://www.denayer.com/download/org.openhab.binding.nikobus-2.5.1-SNAPSHOT.jar

### CODE REVIEW

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusBaseThingHandler.java [hunk 1]
```diff
@@ -20,15 +20,21 @@
...
  * The {@link NikobusBaseThingHandler} class defines utility logic to be consumed by Nikobus thing(s).
  *
  * @author Boris Krivonog - Initial contribution
+ * @author Wouter Denayer - support for module addresses as seen in the Niko PC tool
  */
 @NonNullByDefault
 abstract class NikobusBaseThingHandler extends BaseThingHandler {
-    private @Nullable String address;
+    @Nullable
+    protected String address;
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusBaseThingHandler.java]: Revert this change. @Nullable should be positioned as was.

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 2]
```diff
@@ -111,7 +113,53 @@ public NikobusPushButtonHandler(Thing thing) {
...
+                addressPCString = addressPCString.substring(0,
+                        addressPCString.length() - addressButtonPCString.length()) + addressButtonPCString;
+
+                // reverse bits
+                StringBuilder sb = new StringBuilder(addressPCString);
+                addressPCString = sb.reverse().toString();
+                logger.trace("address Niko button '{}'", addressPCString);
+
+                address = String.format("%06X", Integer.parseInt(addressPCString, 2));
+                // logger.debug("addressNiko '{}'", address);
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java]: Removed commented out code.

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 3]
```diff
@@ -35,14 +36,63 @@
 		</channels>
 
 		<config-description>
-			<parameter name="address" type="text">
+			<parameter-group name="parameterGroupPC">
+				<label>Parameters retrieved from the Niko PC application</label>
```
Comment (Hilbrand) [inline bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: Labels should be short and should follow the convention related to use of uppercase first characters as described in: https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions

Also check the other labels.

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java (hunk 2 above)
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java]: ok

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusBaseThingHandler.java (hunk 1 above)
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusBaseThingHandler.java]: ok

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 3 above)
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: Q: for parameter group labels the longer <description> doesn't seem visible in Paper UI. A shortened label may not be that clear to users + there is enough of room for the label to be shown. How short should it be?

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusBaseThingHandler.java [hunk 4]
```diff
@@ -38,8 +44,15 @@ protected NikobusBaseThingHandler(Thing thing) {
     public void initialize() {
         address = (String) getConfig().get(NikobusBindingConstants.CONFIG_ADDRESS);
         if (address == null) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.CONFIGURATION_ERROR, "Address must be set!");
-            return;
+            String addressPC = (String) getConfig().get(NikobusBindingConstants.CONFIG_ADDRESS_PC);
+            if (addressPC != null) {
+                address = addressPC.substring(2, 4) + addressPC.substring(0, 2);
```
Comment (crnjan) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusBaseThingHandler.java]: would suggest to make an abstract function that does the parsing of `CONFIG_ADDRESS_PC` 
i.e. 
`address = parseNikobusAddress(addressPC)`
where `parseNikobusAddress` is an abstract method and implement it in
1. `NikobusModuleHandler` for modules (actually move this code from `NikobusBaseThingHandler` to `NikobusModuleHandler`) since its same for all modules;
2. in `NikobusPushButtonHandler` for buttons (it's there already, just move it to `parseNikobusAddress`)

That way we would have the code where it belongs.

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 5]
```diff
@@ -35,14 +36,63 @@
...
+				<label>Specific Button</label>
+				<description>The specific button</description>
+				<options>
+					<option value="2_1">2-button A (top)</option>
+					<option value="2_3">2-button B (bottom)</option>
+					<option value="4_1">4-button A (top left)</option>
+					<option value="4_3">4-button B (bottom left)</option>
+					<option value="4_0">4-button C (top right)</option>
+					<option value="4_2">4-button D (bottom right)</option>
+					<option value="8_5">8-button 1A (top left)</option>
```
Comment (crnjan) [inline bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: To be honest, this does not feel very natural - if the goal is to have same config in Nikobus PC software and in OH, than I would suggest to use 1:1 mapping - please note many people still use text configurations (and not i.e. PaperUI) and at least I would need to check what `8_5` actually maps to.

That said I would suggest to use something similar to `"<address>:<number of buttons>:<button>"`

this way we are using the well established OH config format using : as separator and unify the address definition for modules and buttons (no need for the additional `addressButtonPC` parameter).

`addressPC="13F588:8:1A"`

Of course, mapping needs to be implemented - so 1A -> 5 in this case. wdyt?

**Comment by crnjan** [comment]
@gitwouter - thx for this addition it really makes configuring things easier. At the beginning I had similar idea to move away from "bus" definition and use addresses from Nikobus PC software - but since Nikobus was already superseded by Niko Home Control I don't expect much new installations so mainly users porting from v1 -> v2 and therefore (IMHO) it makes more sense to keep addresses as they are to make transition easier. But having the option to use both is even better ;)

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 5 above)
Comment (crnjan) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: Is number of buttons actually needed?

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 6]
```diff
@@ -35,14 +36,63 @@
...
-			<parameter name="address" type="text">
+			<parameter-group name="parameterGroupPC">
+				<label>Parameters retrieved from the Niko PC application</label>
+				<description>Parameters retrieved from the Niko PC application</description>
+				<advanced>false</advanced>
+			</parameter-group>
+			<parameter-group name="parameterGroupBus">
+				<label>Parameters retrieved from the bus</label>
+				<description>String</description>
+				<advanced>true</advanced>
```
[the remark this answers is not in the document]
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: ok

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 7]
```diff
@@ -120,26 +168,35 @@ public void initialize() {
...
-            for (String impactedModuleString : impactedModulesString) {
-                ImpactedModuleUID impactedModuleUID = new ImpactedModuleUID(impactedModuleString.trim());
-                ThingTypeUID thingTypeUID = new ThingTypeUID(bridgeUID.getBindingId(),
-                        impactedModuleUID.getThingTypeId());
-                ThingUID thingUID = new ThingUID(thingTypeUID, bridgeUID, impactedModuleUID.getThingId());
-                impactedModules.add(new ImpactedModule(thingUID, impactedModuleUID.getGroup()));
+            String impactedModulesStringFull = (String) getConfig().get(CONFIG_IMPACTED_MODULES);
+            if (impactedModulesStringFull != null) {
+
+                String[] impactedModulesString = impactedModulesStringFull.toString().split(",");
```
[the remark this answers is not in the document]
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java]: looks like it, will remove

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 5 above)
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: I agree on the format, addressPC="13F588:8:1A" is more natural (I like PaperUI but understand most people seem to be using the text config). 
I added the number of buttons to the value because the HTML dropdown list in PaperUI didn't like to have multiple entries with the same value.

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 3 above)
Comment (Hilbrand) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: I don't know how short it should be, but it should roughly be around 3 or 4 words. PaperUI is only 1 UI and other UIs might have less space. PaperUI is also being replaced so other or future UI's might show the description (which I think is should be the case and thus is a missing feature in PaperUI) I would not make the decision on how it currently looks in PaperUI.

Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: ok, make sense, thx

**Comment by crnjan** [comment]
@gitwouter - a gentle nudge - can you please let me know what is the status of this PR? Asking since this would be a great contribution, specially because I see people actually installing v2 Nikobus binding and not just migrating from v1 version of the binding and would help configure the binding in a much easier way - thx!

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 8]
```diff
@@ -35,14 +36,63 @@
...
+
+			<parameter name="address" type="text"
+				groupName="parameterGroupBus" pattern="([0-9a-fA-F]){6}">
 				<label>Address</label>
 				<description>The Nikobus address of the module</description>
 			</parameter>
-			<parameter name="impactedModules" type="text">
+			<parameter name="impactedModules" type="text"
+				groupName="parameterGroupBus"
+				pattern="([0-9a-fA-F]){4}-\d(([0-9a-fA-F]){4}-\d)*">
```
[the remark this answers is not in the document]
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: what would one enter for moduleType in the string?

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 5 above)
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: I'm going with 13F588:1A now and have removed the drop down list

**Comment by gitwouter** [comment]
was busy with work, will try to resubmit soon.

**Comment by gitwouter** [comment]
@crnjan could you please give me advice on how to handle the  required '@NonNull ... errors?

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/utils/Utils.java [hunk 9]
```diff
@@ -12,17 +12,52 @@
...
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
 /**
  * The {@link Utils} class defines commonly used utility functions.
  *
  * @author Boris Krivonog - Initial contribution
  */
+@NonNullByDefault
 public class Utils {
     public static void cancel(Future<?> future) {
```
Comment (cpmeister) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/utils/Utils.java]: The error you are seeing in the build are mostly referring to this method. Since you added `@NonNullByDefault` to this class it will assume that all fields and parameters are non-null.
This `cancel` method has now been changed to enforce that the `future` parameter is non-null, so any callers of this method must also make sure that they are passing non-null futures. 
It is apparent from the implementation of this method that the `future` parameter was intended to be nullable, as were all the callers of this method assumed it could accept null as well.
In short, to fix the errors just make the parameter nullable.
```suggestion
    public static void cancel(@Nullable Future<?> future) {
```

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java [hunk 10]
```diff
@@ -159,9 +160,20 @@ private void processReceivedValue(byte value) {
...
     @SuppressWarnings("null")
     public void addListener(String command, Runnable listener) {
+        logger.trace("adding listener for '{}'", command);
         if (commandListeners.put(command, listener) != null) {
             logger.warn("Multiple registrations for '{}'", command);
         }
+        
+        // the connection must be open to be able to receive messages from the bus
+        try {
+            connectIfNeeded(connection);
```
Comment (crnjan) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java]: Hm is this really required here? connection will be establish when first command is send (`REFRESH` for example) ... so I guess not need to call it here too, or did I miss something?

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 11]
```diff
@@ -42,13 +46,16 @@
...
+ * buttons.
  *
  * @author Boris Krivonog - Initial contribution
- * @author Wouter Denayer - support for module addresses as seen in the Niko PC tool
+ * @author Wouter Denayer - support for module addresses as seen in Niko PC app
+ * 
  */
 @NonNullByDefault
 public class NikobusPushButtonHandler extends NikobusBaseThingHandler {
+    @NonNullByDefault
```
Comment (crnjan) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java]: AFAIK this is redundand since `NikobusPushButtonHandler` is already marked with  `@NonNullByDefault`

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 12]
```diff
@@ -234,13 +264,43 @@ public void handleCommand(ChannelUID channelUID, Command command) {
         }
     }
 
+    // called when there is a message on the Nikobus
     private void commandReceived() {
-        if (thing.getStatus() != ThingStatus.ONLINE) {
-            updateStatus(ThingStatus.ONLINE);
-        }
 
+        //logger.debug("button press {}", this.address);
```
Comment (crnjan) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java]: probably can remove the comment out code ...

**Comment by crnjan** [comment]
@gitwouter - did a fast check and added some small observations, will try to make a real review soon(ish) + check the the error you are getting - sorry it's taking soo much time.

Really like the new definition

`Thing push-button pb1 [ addressPC = "28092A:1A" ]`

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java (hunk 10 above)
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java:170]: it is needed, when the binding starts up the connection is not opened and therefore it is not reading bus messages coming from physical buttons.

Comment (crnjan) [reply bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java:170]: When binding starts up, it will schedule `refresh` in it's `initialize` function - and that will establish the connection. If that would not be the case then binding would not work at all, right?

The only scenario I can think of that connection would not be established is if there are no modules defined for a bridge - than binding will not try to connect since there is no Nikobus module configured. Are you using a configuration without any Nikobus module (switch/dimmer/rollershutter)?

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 13]
```diff
@@ -219,4 +333,17 @@ private void update() {
     protected String getAddress() {
         return "#N" + super.getAddress();
     }
+
+    private Integer getButtonIndex(String buttonName) {
+        Integer index = buttonIndexMap.get(buttonName.trim());
+
+        //logger.trace("getButtonIndex '{}' '{}'", buttonName, index);
```
Comment (cpmeister) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java:340]: use it or lose it

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml [hunk 14]
```diff
@@ -35,14 +36,63 @@
...
+					<option value="8_4">8-button 1C (top column 2)</option>
+					<option value="8_6">8-button 1D (bottom column 2)</option>
+					<option value="8_1">8-button 2A (top column 3)</option>
+					<option value="8_3">8-button 2B (bottom column 3)</option>
+					<option value="8_0">8-button 2C (top right)</option>
+					<option value="8_2">8-button 2D (bottom right)</option>
+				</options>
+				<default></default>
+			</parameter>
+			<parameter name="impactedModulesPC" type="text"
```
[the remark this answers is not in the document]
Comment (cpmeister) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml:55]: Is this supposed to be removed?

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 15]
```diff
@@ -177,13 +263,41 @@ public void handleCommand(ChannelUID channelUID, Command command) {
...
+            addressPCString = sb.reverse().toString();
+            // last three bits hold the specific button values (0-7)
+            int lastThreeBits = Integer.parseInt(addressPCString.substring(addressPCString.length() - 3), 2);
+            // transform into the letter code (A, B, 1A, ...)
+            String buttonName = Utils.getKeyByValue(buttonIndexMap, lastThreeBits);
+            // drop the last two bits, then set the last bit to 0 to get the base address
+            String baseAddress = addressPCString.substring(0, addressPCString.length() - 3) + "0";
+            int baseAddressInt = Integer.parseInt(baseAddress, 2);
+            logger.trace(" button press bus {}, pc {}:{}", this.address, String.format("%06X", baseAddressInt), buttonName);
+
```
Comment (cpmeister) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java:293]: remove blank line

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 16]
```diff
@@ -120,30 +154,82 @@ public void initialize() {
...
-            String[] impactedModulesString = getConfig().get(CONFIG_IMPACTED_MODULES).toString().split(",");
-            for (String impactedModuleString : impactedModulesString) {
-                ImpactedModuleUID impactedModuleUID = new ImpactedModuleUID(impactedModuleString.trim());
-                ThingTypeUID thingTypeUID = new ThingTypeUID(bridgeUID.getBindingId(),
-                        impactedModuleUID.getThingTypeId());
-                ThingUID thingUID = new ThingUID(thingTypeUID, bridgeUID, impactedModuleUID.getThingId());
-                impactedModules.add(new ImpactedModule(thingUID, impactedModuleUID.getGroup()));
+            String impactedModulesStringFull = (String) getConfig().get(modulesConfigParameter);
+            if (impactedModulesStringFull != null) {
+
```
Comment (cpmeister) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java:165]: remove blank line

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 17]
```diff
@@ -120,30 +154,82 @@ public void initialize() {
...
+                            impactedModuleUID.getThingTypeId());
+                    ThingUID thingUID = new ThingUID(thingTypeUID, bridgeUID, impactedModuleUID.getThingId());
+                    impactedModules.add(new ImpactedModule(thingUID, impactedModuleUID.getGroup()));
+                }
+
+                logger.debug("Impacted modules for {} = {}", thing.getUID(), impactedModules);
+            } else {
+                logger.debug("Impacted modules for {} = none specified", thing.getUID());
             }
+
```
Comment (cpmeister) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java:180]: remove blank line

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 18]
```diff
@@ -104,14 +113,39 @@ protected int getMinimalNumberOfSegments() {
...
     private @Nullable Future<?> requestUpdateFuture;
+    private Map<String, Integer> buttonIndexMap;
 
     public NikobusPushButtonHandler(Thing thing) {
         super(thing);
+
+        buttonIndexMap = Stream
+                .of(new Object[][] { { "A", 1 }, { "B", 3 }, { "C", 0 }, { "D", 2 }, { "1A", 5 }, { "1B", 7 },
+                        { "1C", 4 }, { "1D", 6 }, { "2A", 1 }, { "2B", 3 }, { "2C", 0 }, { "2D", 2 }, })
+                .collect(Collectors.toMap(p -> (String) p[0], p -> (Integer) p[1]));
```
Comment (cpmeister) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java:124]: Just populate the map usual way, this way isn't easier to read and is much more prone to developer error.

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java [hunk 19]
```diff
@@ -159,9 +160,20 @@ private void processReceivedValue(byte value) {
...
+        } catch (IOException e) {
+            logger.debug("Sending command failed due {}", e.getMessage(), e);
+            connection.close();
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+
```
Comment (cpmeister) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java:175]: remove blank line

File: bundles/org.openhab.binding.nikobus/README.md [hunk 20]
```diff
@@ -174,6 +227,19 @@ Thing push-button pb1 [ address = "28092A", impactedModules = "switch-module:s1:
 
 In addition to the status requests triggered by button presses, there is also a scheduled status update interval defined by the `refreshInterval` parameter and explained above.
 
+### Rules
+
+When a Thing has been created for a button it is also possible to trigger a rule when the physical button is pressed. In the below example the rule is triggered when the Button Event Channel `nikobus:push-button:b6e47184:button` fires. An `ON` command is then sent to the kitchen light.
```
Comment (cpmeister) [inline bundles/org.openhab.binding.nikobus/README.md:232]: Please put each sentence on its own line in addition to these changes:
```suggestion
When a Thing has been created for a button it is also possible to trigger a rule when the physical button is pressed. In the example below, the rule is triggered when the Button Event Channel `nikobus:push-button:b6e47184:button` fires. An `ON` command is then sent to the kitchen light.
```

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java (hunk 15 above)
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java:293]: I need impactedModulesPC for display in PaperUI

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java (hunk 10 above)
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java:170]: I am slowly moving a way from Nikobus. My modules are failing one by one and soon there will be no spare parts. But I want to keep the physical buttons in my house. So I would like the binding to work even without any modules defined.

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 14 above)
[the remark this answers is not in the document]
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml:55]: I need it for Paper UI

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusBaseThingHandler.java (hunk 4 above)
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusBaseThingHandler.java]: done

File: bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml (hunk 5 above)
Comment (gitwouter) [reply bundles/org.openhab.binding.nikobus/src/main/resources/ESH-INF/thing/thing-types.xml]: done

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java (hunk 10 above)
Comment (crnjan) [reply bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPcLinkHandler.java:170]: That is not good to hear, hope my modules will last a bit longer :)

In that case (setup without modules) I would suggest to move this `connect` related code to `refresh` - please note `addListener` will be most probably called a lot of times and no real need to request connection here.

So I would suggest to modify 

```
if (things.isEmpty()) {
            logger.debug("Nothing to refresh");
            return;
        }
```

within the `refresh` method:

```
if (things.isEmpty()) {
  if (!commandListeners.isEmpty()) {
       // the connection must be open to be able to receive messages from the bus
        try {
            connectIfNeeded(connection);
         } catch (IOException e) {
            logger.debug("Sending command failed due {}", e.getMessage(), e);
            connection.close();
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
  }
 logger.debug("Nothing to refresh");
 return;
 }
```

Makes sense?

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 21]
```diff
@@ -120,30 +154,82 @@ public void initialize() {
...
         }
 
-        logger.debug("Impacted modules for {} = {}", thing.getUID(), impactedModules);
-
         NikobusPcLinkHandler pcLink = getPcLink();
         if (pcLink != null) {
             pcLink.addListener(getAddress(), this::commandReceived);
         }
+
+        updateStatus(ThingStatus.ONLINE);
```
Comment (crnjan) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java:191]: Why is button put online here already?

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java [hunk 22]
```diff
@@ -177,13 +263,41 @@ public void handleCommand(ChannelUID channelUID, Command command) {
         }
     }
 
+    // called when there is a message on the Nikobus
     private void commandReceived() {
-        if (thing.getStatus() != ThingStatus.ONLINE) {
```
Comment (crnjan) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/handler/NikobusPushButtonHandler.java:181]: Button is put online when first Nikobus command is received - meaning the button is mapped correctly. AFAIK there is no other way to check if configured address is correct.

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/protocol/NikobusCommand.java [hunk 23]
```diff
@@ -28,6 +28,7 @@
  */
 @NonNullByDefault
 public class NikobusCommand {
+    @NonNullByDefault
```
Comment (crnjan) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/protocol/NikobusCommand.java:31]: Redundand AFAIK

File: bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/protocol/NikobusCommand.java [hunk 24]
```diff
@@ -46,6 +47,7 @@ public String get() throws Exception {
         }
     }
 
+    @NonNullByDefault
```
Comment (crnjan) [inline bundles/org.openhab.binding.nikobus/src/main/java/org/openhab/binding/nikobus/internal/protocol/NikobusCommand.java:50]: Redundand AFAIK
